### PR TITLE
CO-2671_EnvSource_doesn_t_with_multi-valued_email_attribute

### PR DESCRIPTION
**Pull request by ioigoume**

### CODE REVIEW

**Comment by boshrin** [comment]
See comment on [CO-2671](https://todos.internet2.edu/browse/CO-2671).

File: app/Plugin/EnvSource/Config/Schema/schema.xml [hunk 1]
```diff
@@ -39,6 +39,7 @@
     </field>
     <field name="duplicate_mode" type="C" size="2" />
     <field name="redirect_on_duplicate" type="C" size="1024" />
+    <field name="sp_type" type="C" size="1" />
```
Comment (boshrin) [inline app/Plugin/EnvSource/Config/Schema/schema.xml:42]: For consistency with other enums, this should be size 2.

File: app/Plugin/EnvSource/Lib/enum.php [hunk 2]
```diff
@@ -31,3 +31,8 @@ class EnvSourceDuplicateModeEnum
   const LoginIdentifier = "LI";
   const SORIdentifier   = "SI";
 }
+
+class AuthProviderEnum
+{
+  const Shibboleth   = ";";
```
Comment (boshrin) [inline app/Plugin/EnvSource/Lib/enum.php:37]: This is going to make SQL query construction really annoying :) The enum const should be something like "SH" (for "Shibboleth").

We should also define "Other" (code "O" for consistency") so if any deployer is using something else they don't get confused. We could proactively define "SimpleSamlPHP" ("SS"?) as well even though we're not doing anything specific for it.

File: app/Plugin/EnvSource/Lib/lang.php [hunk 3]
```diff
@@ -42,7 +42,11 @@
     EnvSourceDuplicateModeEnum::AnyIdentifier    => 'Any Identifier Match',
     EnvSourceDuplicateModeEnum::LoginIdentifier  => 'Login Identifier Match',
   ),
-  
+
+  'pl.envsource.en.auth.provider' => array(
+    AuthProviderEnum::Shibboleth    => 'Shibboleth',
```
Comment (boshrin) [inline app/Plugin/EnvSource/Lib/lang.php:47]: For pedantic reasons, this should probably say "Shibboleth SP".

File: app/Plugin/EnvSource/Lib/lang.php [hunk 4]
```diff
@@ -52,8 +56,9 @@
   'er.envsource.token'          => 'Token error',
   
   // Plugin texts
-  'pl.envsource.mode.dupe'      => 'Duplicate Handling Mode',
-  'pl.envsource.name.unknown'   => 'Unknownname',
-  'pl.envsource.redirect.dupe'  => 'Redirect on Duplicate',
-  'pl.envsource.sorid.desc'     => 'This must be set to an environment variable holding a unique identifier for any authenticated user.'
+  'pl.envsource.mode.dupe'               => 'Duplicate Handling Mode',
+  'pl.envsource.sp.type'                 => 'Multi Value Delimiter',
```
Comment (boshrin) [inline app/Plugin/EnvSource/Lib/lang.php:60]: The goal here is to document which SP is used for the integration, not the behavior of that SP. So this field should be labeled something like "Web Server SP Provider".

**Review by boshrin — CHANGES_REQUESTED** [review]

File: app/Plugin/EnvSource/Config/Schema/schema.xml (hunk 1 above)
Comment (ioigoume) [reply app/Plugin/EnvSource/Config/Schema/schema.xml:42]: Done

File: app/Plugin/EnvSource/Lib/lang.php (hunk 3 above)
Comment (ioigoume) [reply app/Plugin/EnvSource/Lib/lang.php:47]: Done

File: app/Plugin/EnvSource/Lib/lang.php (hunk 4 above)
Comment (ioigoume) [reply app/Plugin/EnvSource/Lib/lang.php:60]: Done

File: app/Plugin/EnvSource/Lib/enum.php (hunk 2 above)
Comment (ioigoume) [reply app/Plugin/EnvSource/Lib/enum.php]: Done

File: app/Plugin/EnvSource/Config/Schema/schema.xml [hunk 5]
```diff
@@ -39,6 +39,8 @@
     </field>
     <field name="duplicate_mode" type="C" size="2" />
     <field name="redirect_on_duplicate" type="C" size="1024" />
+    <field name="sp_type" type="C" size="2" />
+    <field name="other_type" type="C" size="2" />
```
Comment (boshrin) [inline app/Plugin/EnvSource/Config/Schema/schema.xml:43]: I don't think we need `other_type`, see additional comment below.

File: app/Plugin/EnvSource/Model/EnvSourceBackend.php [hunk 6]
```diff
@@ -126,9 +126,34 @@ protected function resultToOrgIdentity($result) {
...
-      $orgdata['EmailAddress'][0]['type'] = EmailAddressEnum::Official;
-      $orgdata['EmailAddress'][0]['verified'] = true;
+      if(!empty($this->pluginCfg['sp_type'])) {
+        $delimiter = "";
+        if($this->pluginCfg['sp_type'] == AuthProviderEnum::Shibboleth) {
+          $delimiter = ";";
+        } elseif($this->pluginCfg['sp_type'] == AuthProviderEnum::Simplesamlphp) {
+          $delimiter =",";
+        } elseif($this->pluginCfg['sp_type'] == AuthProviderEnum::Other) {
+          $delimiter =  trim($this->pluginCfg['other_type']);
```
Comment (boshrin) [inline app/Plugin/EnvSource/Model/EnvSourceBackend.php:136]: If type is `Other` we simply don't support multiple values, and we don't try to `explode()` the value. This also simplifies the user interface, below.

**Review by boshrin — CHANGES_REQUESTED** [review]

File: app/Plugin/EnvSource/Model/EnvSourceBackend.php (hunk 6 above)
Comment (ioigoume) [reply app/Plugin/EnvSource/Model/EnvSourceBackend.php:136]: Done

File: app/Plugin/EnvSource/Config/Schema/schema.xml (hunk 5 above)
Comment (ioigoume) [reply app/Plugin/EnvSource/Config/Schema/schema.xml:43]: Done

**Pull request merged**